ibis/backends/duckdb/compiler/core.py

```diff
@@ -0,0 +1,99 @@
+"""DuckDB ibis expression to sqlglot compiler.
+
+The compiler is built with a few `singledispatch` functions:
+
+    1. `translate` for table expressions
+    1. `translate` for table nodes
+    1. `translate_rel`
+    1. `translate_val`
+
+## `translate`
+
+### Expression Implementation
+
+The table expression implementation of `translate` is a pass through to the
+node implementation.
+
+### Node Implementation
+
+There's a single `ops.Node` implementation for `ops.TableNode`s instances.
+
+This function:
+
+    1. Topologically sorts the expression graph.
+    1. Seeds the compilation cache with in-degree-zero table names.
+    1. Iterates though nodes with at least one in-degree and places the result
+       in the compilation cache. The cache is used to construct `ops.TableNode`
+       keyword arguments to the current translation rule.
+
+## `translate_rel`
+
+Translates a table operation given already-translated table inputs.
+
+If a table node needs to translate value expressions, for example, an
+`ops.Aggregation` that rule is responsible for calling `translate_val`.
+
+## `translate_val`
+
+Recurses top-down and translates the arguments of the value expression and uses
+those as input to construct the output.
+"""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any
+
+import sqlglot as sg
+
+import ibis.expr.operations as ops
+import ibis.expr.types as ir
+from ibis.backends.duckdb.compiler.relations import translate_rel
+
+if TYPE_CHECKING:
+    from collections.abc import Mapping
+
+
+def translate(op: ops.TableNode, params: Mapping[ir.Value, Any]) -> sg.exp.Expression:
```

Comment:
A follow-up is to move this into `ibis/backends/base/sqlglot/core.py` and reuse it for both the clickhouse and duckdb backends.

I think we may we want to extract subqueries before doing that though.

I did play around with using sqlglot's optimizer, which can remove duplicate subqueries, but it's also defeated by a bunch of `dot_sql` tests and functionality.